backintime-library/runtime/src/commonMain/kotlin/com/github/kitakkun/backintime/runtime/BackInTimeDebugService.kt

```diff
@@ -19,55 +22,115 @@ import kotlin.coroutines.CoroutineContext
 
-    private val instances = WeakHashMap<BackInTimeDebuggable, String>()
+    private val instances = mutableMapOf<String, WeakReference<BackInTimeDebuggable>>()
+    private val serviceEventDispatchQueue = mutableListOf<BackInTimeDebugServiceEvent>()
 
-    private val mutableServiceEventFlow = MutableSharedFlow<BackInTimeDebugServiceEvent>()
-    val serviceEventFlow = mutableServiceEventFlow.asSharedFlow()
+    private var connector: BackInTimeConnector? = null
+    private var processDebuggerEventJob: Job? = null
+    private var observeConnectedFlowJob: Job? = null
 
-    private val mutableInternalErrorFlow = MutableSharedFlow<Throwable>()
-    val internalErrorFlow = mutableInternalErrorFlow.asSharedFlow()
+    init {
+        // clean up instances that are garbage collected
+        startInstanceCleanUpJob()
+    }
 
-    private val internalServiceEventQueue = mutableListOf<BackInTimeDebugServiceEvent>()
-    private var isConnected: Boolean = false
+    fun setConnector(connector: BackInTimeConnector) {
+        this.connector = connector
+    }
 
-    /**
-     * start processing events (also process queued events)
-     * should be called from FlipperPlugin
-     */
-    fun start() {
-        isConnected = true
-        internalServiceEventQueue.forEach { event -> launch { mutableServiceEventFlow.emit(event) } }
-        internalServiceEventQueue.clear()
+    fun startService() {
+        val connector = this.connector ?: return
+        connector.connect()
+        processDebuggerEventJob = launch {
+            connector.receiveEventAsFlow().collect { event ->
+                val result = processDebuggerRequest(event) ?: return@collect
+                sendOrQueue(result)
+            }
+        }
+        observeConnectedFlowJob = launch {
+            connector.connectedFlow.filter { it }.collect {
+                serviceEventDispatchQueue.forEach { event ->
+                    connector.sendEvent(event)
+                }
+                serviceEventDispatchQueue.clear()
+            }
+        }
+    }
+
+    fun stopService() {
+        processDebuggerEventJob?.cancel()
+        observeConnectedFlowJob?.cancel()
+        connector?.disconnect()
+        processDebuggerEventJob = null
+        observeConnectedFlowJob = null
+        connector = null
+    }
+
+    private fun startInstanceCleanUpJob() {
+        launch {
+            while (true) {
+                instances.keys.forEach { key ->
+                    val instance = instances[key] ?: return@forEach
+                    if (instance.get() == null) {
+                        instances.remove(key)
+                    }
+                }
+                delay(1000 * 15)
+            }
+        }
+    }
+
+    private fun sendOrQueue(event: BackInTimeDebugServiceEvent) {
+        if (connector?.isConnected == true) {
+            connector?.sendEvent(event)
+        } else {
+            serviceEventDispatchQueue.add(event)
+        }
     }
```

Comment:
when it has an active connection, directly send the event.
when it has no connection, just push it to the queue.